File: cap.cc
#include "cap.h"
#include <pylon/PylonImage.h>
#include "opencv2/core/core.hpp"

using namespace std;

using namespace Pylon;

Capture::Capture(std::string serial)
{
  try
  {
    CTlFactory& tlFactory = CTlFactory::GetInstance();
    DeviceInfoList_t devices;
    tlFactory.EnumerateDevices(devices);
    for (auto& device : devices)
      {
	if (device.GetSerialNumber() == String_t(serial.c_str()))
	  {
	    camera.Attach(tlFactory.CreateDevice(device));
	  }
      }
    camera.Open();

    cout << "Using device " << camera.GetDeviceInfo().GetModelName() << endl;
    converter.OutputPixelFormat = PixelType_BGR8packed;
    converter.OutputBitAlignment = OutputBitAlignment_MsbAligned;
  }
  catch (GenICam::GenericException & e)
  {
    //camera = nullptr;
    cerr << "An exception occurred" << endl;
    cout << e.GetDescription() << endl;
    exit(1);
  }
}

void Capture::start()
{
  GenApi::INodeMap& nodemap = camera.GetNodeMap();
  CEnumParameter gainAuto(nodemap, "GainAuto");
  gainAuto.TrySetValue("Continuous");
  camera.StartGrabbing(Pylon::GrabStrategy_LatestImageOnly);
}

void Capture::grab(cv::Mat& img)
{
  if (!camera.IsGrabbing())
    {
      cout << "Error: Not grabbing" << endl;
      return;
    }
  CPylonImage image;
  camera.RetrieveResult(1000, ptrGrabResult, TimeoutHandling_ThrowException);
  if (ptrGrabResult->GrabSucceeded())
    {
      converter.Convert(image, ptrGrabResult);

      img = cv::Mat(ptrGrabResult->GetHeight(), ptrGrabResult->GetWidth(), CV_8UC3);
      memcpy(img.ptr(), image.GetBuffer(), ptrGrabResult->GetHeight() * ptrGrabResult->GetWidth() * 3);
      /*
      cv::namedWindow("TEST", cv::WINDOW_NORMAL);
      cv::imshow("TEST", img);
      cv::waitKey(3000);*/
    }
}

void Capture::stop()
{
  camera.StopGrabbing();
  camera.Close();
}
/*
int main()
{
  cv::Mat img;
  if (img.empty())
    {
      cout << "Image empty at start" << endl;
    }
  
  Pylon::PylonAutoInitTerm autoInitTerm;
  
  Capture cap("40016577");
  cap.start();
  cap.grab(img);

  if (img.empty())
    {
      cout << "Empty frame" << endl;
    }

  cout << "Size: " << img.size << endl;

  //cv::imwrite("test_IMAGE.png", img);
  
  cv::Mat hsv;
  //cv::cvtColor(img, hsv, cv::COLOR_BGR2HSV);
  
  cv::namedWindow("test", cv::WINDOW_NORMAL);
  cout << "Named window" << endl;
  cv::imshow("test", img);
  cout << "Imshow" << endl;
  cv::waitKey(5000);

  cap.stop();

  return 0;
}
*/

File: cap.py
# cap.py

# Defines the Capture class used to access a camera
# Always accesses the first camera found

import numpy as np
import cv2 as cv
import time
from pypylon import pylon
from pypylon import genicam


class Capture(object):
    def __init__(self):
        self.camera = None
        self.converter = pylon.ImageFormatConverter()
        self.converter.OutputPixelFormat = pylon.PixelType_BGR8packed
        self.converter.OutputBitAlignment = pylon.OutputBitAlignment_MsbAligned
        self.connect()

    def connect(self):
        try:
            self.camera = pylon.InstantCamera(pylon.TlFactory.GetInstance().CreateFirstDevice())
            self.camera.Open()
            print("Connecting to camera:", self.camera.GetDeviceInfo().GetModelName())
        except genicam.GenericException as e:
            self.camera = None
            pass

    def start(self):
        if self.camera is None:
            print("Error: Not connected to camera, can't start grabbing.")
            return self
        self.camera.StartGrabbing(pylon.GrabStrategy_LatestImageOnly)
        self.set_param('GainAuto', 'Continuous')
        self.set_param('AcquisitionFrameRateEnable', True)
        self.set_param('AcquisitionFrameRateAbs', 20.0)
        return self

    def grab(self):
        if self.camera is None or not self.camera.IsGrabbing():
            print("Error: Not connected to camera or not grabbing")
            return None
        result = self.camera.RetrieveResult(1000, pylon.TimeoutHandling_ThrowException)

        img = None
        if result.GrabSucceeded():
            image = self.converter.Convert(result)
            img = image.GetArray()

        result.Release()
        return img

    def stop(self):
        if self.camera is None:
            print("Error: Camera not started, can't stop.")
            return self
        self.camera.StopGrabbing()
        self.camera.Close()
        self.camera = None
        return self

    def set_param(self, param, value):
        if self.camera is None:
            print("Error: Not connected to any camera.")
            return self
        setattr(self.camera, param, value)
        return self

    def get_param(self, param):
        if self.camera is None:
            print("Error: Not connected to any camera.")
            return None
        val = getattr(self.camera, param)
        return val.GetValue()

if __name__ == '__main__':
    cap = Capture()
    width = cap.get_param("Width")
    height = cap.get_param("Height")
    print("Image resolution: " + str(width) + ", " + str(height))
    cap.start()
    i = 0
    while True:
        im = cap.grab()
        if im is None:
            print("Error: Got empty result.")
            break
        cv.namedWindow('cap', cv.WINDOW_NORMAL)
        cv.imshow('cap', im)
        ch = cv.waitKey(1)
        if ch == 27 or ch == ord('q'):
            break
        if ch == ord('s'):
            temp = cv.selectROI(im)
            cv.imwrite('object.png',im)
            break

    cap.stop()
    cv.destroyAllWindows()
